Quarkus kiota post (#1885)
quarkusio/quarkusio.github.io: _posts/2024-XX-XX-quarkus-kiota.adoc

---
layout: post

> title: 'The road to generated SDKs with Kiota lands to Quarkus'

Reviewer: ```suggestion
title: 'The road to generated SDKs with Kiota using Quarkus'
```

date: 2024-XX-XX
tags: release
synopsis: 'Introducing a Quarkus-Kiota extension and available Java tools to work with Kiota.'
author: andreaTP
---

== The Challenge
:imagesdir: assets/images/posts/kiota

> Quarkus applications typically expose functionalities through API endpoints; here, we discuss how to easily consume those APIs and provide a smooth experience for our beloved users. While HTTP calls are easy to perform in any programming language and environment, when API complexity increases, and the number of available endpoints becomes uncountable, the need for better tools becomes apparent.

Reviewer: nothing wrong in here but supposedly Quarkus already has solutions in the space of handling http client and server interactions. Kiota's "force" is the cross-language uniform support so maybe incorporate that earlier in here?

image::api_monster.jpeg[API monster]

=== Developer Experience Wishlist:

> - Fully typed programmatic APIs that make discovering and invoking HTTP endpoints safe, expressive, and elegant

Reviewer: I know at the end you mention the existing openapi generator - but I do feel since this is on the quarkus main blog it would be good to highlight that Quarkus already has a http client approach with resteasy client that standalone or combined with the open API generator lets you have all these things....and then explain what kiota solves separately (multilanguage variations and potentially dependent on perspective a better code generation approach)

- An easy and idiomatic way to plug in various authentication mechanisms
- Smooth upgrades of the generated code when the description evolves

Looking at the learnings from big companies facing this issue at scale, a viable solution is to ship language-specific SDKs (Software Development Kits). For years, companies have been paying developers to manually provide the abstraction layer, making cloud services functionalities more easily accessible. This approach adds up to the maintenance burden, increases the operational complexity of releases, and, worst, is tedious for developers.

To pave the way for better tooling, new standards are being defined in the computer industry. For building HTTP-based APIs, link:https://www.openapis.org[OpenAPI] stands out as a widely adopted option.
In the context of this article we are going to assume you already have an OpenAPI description of the service.

> === A Story of Open Source Collaboration

Reviewer: I love both Microsoft and Red Hat - but they are big corporations and its not like all of red hat and all of Microsoft did this  - could you highlight bit more specifically what parts of those companies got involved ?

==== Microsoft Kiota

Recognizing the challenges faced by the industry in generating comprehensive and efficient SDKs for diverse APIs, Microsoft took the lead. In the context of link:https://learn.microsoft.com/en-us/graph/overview[Microsoft Graph], they introduced link:https://github.com/microsoft/kiota[Kiota], a CLI to streamline and automate the process of creating Software Development Kits (SDKs) for HTTP APIs.

Kiota, born and raised in the open source, fosters collaboration and is ground-up designed with link:https://learn.microsoft.com/en-gb/openapi/kiota/design[modularity] in mind. The Kiota project's community is extra welcoming, and it has demonstrated to be easy to open an issue not only for bugs but also to improve over the design decisions.

> Nowadays, Kiota demonstrates its value daily with the automated generation of SDKs for an incredible number of Microsoft services. More big industries, like link:https://github.blog/2024-01-03-our-move-to-generated-sdks/[GitHub], are quickly following the lead.

Reviewer: interesting that github is moving to kiota based sdk's ....would be awesome if their setup would be available so java variants could be generated?

Author: The projects are open source, e.g. https://github.com/octokit/go-sdk

Reviewer: thats the generated output is it not ? trying to spot the kiota generation steps...but my quick'n'dirty go-fu is failing me.

Author: It should be in this repo: https://github.com/octokit/source-generator

==== Red Hat

> Deeply understanding the technological problem and accepting the fact that the available alternatives are sub-optimal, especially for consistency across languages and frameworks, Red Hat joined the effort on Kiota. The collaboration led to important milestones, directly blossomed from this collaboration:

Reviewer: ```suggestion
Deeply understanding the technological problem and accepting the fact that the available alternatives are sub-optimal, especially for consistency across languages and frameworks, Apicurio team from Red Hat joined the effort on Kiota. The collaboration led to important milestones, directly blossomed from this collaboration:
```

- To make the usage of the generated code almost indistinguishable from a human-crafted product, we thoroughly implemented language-specific mangling of names.
- To ease the fruition of the Kiota CLI from standard Java projects, we implemented a link:https://github.com/kiota-community/kiota-java-extra?tab=readme-ov-file#maven-plugin[Maven plugin].

> - To make frictionless the usage of Kiota on different and opinionated technological stacks, we ship alternative core libraries to let the user easily swap out from Microsoft's defaults (OkHttp + Gson): link:https://github.com/kiota-community/kiota-java-extra?tab=readme-ov-file#serialization-jackson[Jackson Serde], link:https://github.com/kiota-community/kiota-java-extra?tab=readme-ov-file#http-vertx[Vert.X Http], link:https://github.com/kiota-community/kiota-java-extra?tab=readme-ov-file#http-jdk[JDK Http].

Reviewer: ```suggestion
- To make frictionless the usage of Kiota on Quarkus, we ship alternative core libraries to let the user easily swap out from Microsoft's defaults (OkHttp + Gson): link:https://github.com/kiota-community/kiota-java-extra?tab=readme-ov-file#serialization-jackson[Jackson Serde], link:https://github.com/kiota-community/kiota-java-extra?tab=readme-ov-file#http-vertx[Vert.X Http], link:https://github.com/kiota-community/kiota-java-extra?tab=readme-ov-file#http-jdk[JDK Http].
```

...is there another stack than Quarkus where those choices makes sense?

Author: It depends on the stack people are using, Jackson is widely used instead of Gson AFAIK, but happy to revisit this.

  <artifactId>microsoft-kiota-serialization-multipart</artifactId>
  <version>${kiota.libs.version}</version>
</dependency>
<dependency>
  <groupId>jakarta.annotation</groupId>
  <artifactId>jakarta.annotation-api</artifactId>
</dependency>
----

Now we need to generate the actual client for our OpenAPI description, to do so, you should drop the OpenAPI file(in ``yaml`` or ``json`` format) in the ``src/main/openapi`` folder of your project.
You are all set to use the client in your application!

[source,java]
----
var client = new ApiClient(new VertXRequestAdapter(vertx));
client.
----

striking ``.`` after client, the code completion of your IDE kicks in and provide you a beautiful, fully typed, builder pattern matching the endopoint descriptions provided in the OpenAPI specification.

image::completion.jpeg[Code Completion]

For example an endpoint definition like https://github.com/Apicurio/apicurio-registry/blob/6882af10e9de8e1d245006db01f039b1fbf6355a/common/src/main/resources/META-INF/openapi-v2.json#L668[this one] nicely unroll in Java as:

[source,java]
----
client
  .groups()
  .byGroupId(groupId)
  .artifacts()
  .byArtifactId(artifactId)
  .meta()
  .get();
----

> == Alternatives

Reviewer: I think it is kinda missing an answer to the question - ok, so this kiota thing now exist - what does it mean for me  as a quarkus user - what should I use it for ? when should i use quarkus rest client vs this and does the kiota approach give us the same level of integration with quarkus tracing,monitoring, failover, security etc.

might not know all the answers to that but lets at least guide users a bit here.

---
layout: post
title: 'The road to generated SDKs with Kiota using Quarkus'
date: 2024-XX-XX
tags: release
synopsis: 'Introducing a Quarkus-Kiota extension and available Java tools to work with Kiota.'
author: andreaTP
---

== The Challenge
:imagesdir: assets/images/posts/kiota

Quarkus applications typically expose functionalities through API endpoints; here, we discuss how to easily consume those APIs and provide a smooth experience for our beloved users. While HTTP calls are easy to perform in any programming language and environment, when API complexity increases, and the number of available endpoints becomes uncountable, the need for better tools becomes apparent.

image::api_monster.jpeg[API monster]

=== Developer Experience Wishlist:

- Fully typed programmatic APIs that make discovering and invoking HTTP endpoints safe, expressive, and elegant
- An easy and idiomatic way to plug in various authentication mechanisms
- Smooth upgrades of the generated code when the description evolves

> Looking at the learnings from big companies facing this issue at scale, a viable solution is to ship language-specific SDKs (Software Development Kits). For years, companies have been paying developers to manually provide the abstraction layer, making cloud services functionalities more easily accessible. This approach adds up to the maintenance burden, increases the operational complexity of releases, and, worst, is tedious for developers.

Reviewer: ```suggestion
Looking at the learnings from big companies facing this issue at scale, a viable solution is to ship multiple language-specific SDKs (Software Development Kits), i.e. for Java, Python, Go, JavaScript etc. For years, companies have been paying developers to manually provide the abstraction layer, making cloud services functionalities more easily accessible. This approach adds up to the maintenance burden, increases the operational complexity of releases, and, worst, is tedious for developers.
```

just to make explicit this is about multiple languages.

=== Quarkus Meet Kiota

Now that all the dots have been connected, the final and crucial step is a seamless integration with supersonic, subatomic applications leveraging Quarkus!
This is the motivation behind the birth of the new ``quarkus-kiota`` extension.
The project is in its early stages, and we encourage you to try it out and provide feedback.

The codebase lives in the link:https://github.com/quarkiverse/quarkus-kiota[Quarkiverse], the project is listed in the link:https://quarkus.io/extensions/io.quarkiverse.kiota/quarkus-kiota/[extensions] and the link:https://docs.quarkiverse.io/quarkus-kiota/dev/index.html[docs] are available at the usual location.

Get started by adding this extension:

----
quarkus ext add io.quarkiverse.kiota:quarkus-kiota
----

> Since it's a code generator extension, you will need to add the `generate-code` goal to the `quarkus-maven-plugin` `executions` section:

Reviewer: ```suggestion
Since it's a code generator extension, you will want to check the `generate-code` goal is present in the `quarkus-maven-plugin` `executions` section. Default Quarkus generated projects have it but custom or older projects might not:
```

  <version>{kiota-java-extra.version}</version>
</dependency>
<dependency>
  <groupId>io.kiota</groupId>
  <artifactId>kiota-serialization-jackson-quarkus</artifactId>
  <version>{kiota-java-extra.version}</version>
</dependency>
<dependency>
  <groupId>com.microsoft.kiota</groupId>
  <artifactId>microsoft-kiota-serialization-text</artifactId>
  <version>${kiota.libs.version}</version>
</dependency>
<dependency>
  <groupId>com.microsoft.kiota</groupId>
  <artifactId>microsoft-kiota-serialization-form</artifactId>
  <version>${kiota.libs.version}</version>
</dependency>
<dependency>
  <groupId>com.microsoft.kiota</groupId>
  <artifactId>microsoft-kiota-serialization-multipart</artifactId>
  <version>${kiota.libs.version}</version>
</dependency>
<dependency>
  <groupId>jakarta.annotation</groupId>
  <artifactId>jakarta.annotation-api</artifactId>
</dependency>
----

Now we need to generate the actual client for our OpenAPI description, to do so, you should drop the OpenAPI file(in ``yaml`` or ``json`` format) in the ``src/main/openapi`` folder of your project.
You are all set to use the client in your application!

[source,java]

> ----

Reviewer: would be good to have the imports visible here - just for full transparency.

Author: 👍

Author: Done here:
https://github.com/quarkusio/quarkusio.github.io/pull/1885/commits/850a8a6a8538f985309786070feba4b0668e2866

More specifically:
`io.apisdk.example.yaml.ApiClient`: is the generated code
`io.kiota.http.vertx.VertXRequestAdapter`: is the `RequestAdapter` implementation we are shipping [here](https://github.com/kiota-community/kiota-java-extra/tree/main?tab=readme-ov-file#http-vertx)

To pave the way for better tooling, new standards are being defined in the computer industry. For building HTTP-based APIs, link:https://www.openapis.org[OpenAPI] stands out as a widely adopted option.
In the context of this article we are going to assume you already have an OpenAPI description of the service.

=== Current Status in Quarkus

Quarkus already provides an expressive and fully integrated link:https://quarkus.io/guides/rest-client-reactive[Rest Client] to enable you to craft HTTP calls beautifully.
Additionally, the link:https://github.com/quarkiverse/quarkus-openapi-generator[quarkus-openapi-generator] is a mature Quarkus extension designed to generate client code from OpenAPI.
This option is recommended when a tight integration with Quarkus is your primary focus.

In this post, we will introduce an alternative solution that addresses the same problem but makes different trade-offs to prioritize consistency across languages and frameworks.

=== A Story of Open Source Collaboration

> ==== Microsoft Graph - Kiota

Reviewer 2: ```suggestion
==== Microsoft - Kiota
```

(this is how we've been introducing the tool to date, I'd make sure the name of the project is consistent, but keep the team name to outline the collaboration work)

Author: @maxandersen specifically asked to define the teams involved in this collaboration.
I'm happy with whatever you two agree on.

Reviewer: changing the section title is fine if ok by Microsoft  - the specific origin is mentioned in the body which is fine.

This is the motivation behind the birth of the new ``quarkus-kiota`` extension.
The project is in its early stages, and we encourage you to try it out and provide feedback.

The codebase lives in the link:https://github.com/quarkiverse/quarkus-kiota[Quarkiverse], the project is listed in the link:https://quarkus.io/extensions/io.quarkiverse.kiota/quarkus-kiota/[extensions] and the link:https://docs.quarkiverse.io/quarkus-kiota/dev/index.html[docs] are available at the usual location.

Get started by adding this extension:

----
quarkus ext add io.quarkiverse.kiota:quarkus-kiota
----

Since it's a code generator extension, you will want to check the `generate-code` goal is present in the `quarkus-maven-plugin` `executions` section. Default Quarkus generated projects have it but custom or older projects might not:

[source,xml]
----
<plugin>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-maven-plugin</artifactId>
  <extensions>true</extensions>
  <executions>
    <execution>
      <goals>
        <goal>build</goal>
        <goal>generate-code</goal>
      </goals>
    </execution>
  </executions>
</plugin>
----

> Add the necessary dependencies, we are going to replace the default http and json serialization libraries published by Microsoft and swap in Vert.X and Jackson(published from this https://github.com/kiota-community/kiota-java-extra[repository]) as they nicely play with the rest of the stack of a typical Quarkus based application.

Reviewer 3: ```suggestion
Add the necessary dependencies, we are going to replace the default HTTP and JSON serialization libraries published by Microsoft and swap in Vert.X and Jackson(published from this https://github.com/kiota-community/kiota-java-extra[repository]) as they nicely play with the rest of the stack of a typical Quarkus based application.
```

----
<dependency>
  <groupId>com.microsoft.kiota</groupId>
  <artifactId>microsoft-kiota-abstractions</artifactId>
  <version>${kiota.libs.version}</version>
</dependency>
<dependency>
  <groupId>io.kiota</groupId>
  <artifactId>kiota-http-vertx</artifactId> <!-- alternatively <artifactId>kiota-http-jdk</artifactId> -->
  <version>{kiota-java-extra.version}</version>
</dependency>
<dependency>
  <groupId>io.kiota</groupId>
  <artifactId>kiota-serialization-jackson-quarkus</artifactId>
  <version>{kiota-java-extra.version}</version>
</dependency>
<dependency>
  <groupId>com.microsoft.kiota</groupId>
  <artifactId>microsoft-kiota-serialization-text</artifactId>
  <version>${kiota.libs.version}</version>
</dependency>
<dependency>
  <groupId>com.microsoft.kiota</groupId>
  <artifactId>microsoft-kiota-serialization-form</artifactId>
  <version>${kiota.libs.version}</version>
</dependency>
<dependency>
  <groupId>com.microsoft.kiota</groupId>
  <artifactId>microsoft-kiota-serialization-multipart</artifactId>
  <version>${kiota.libs.version}</version>
</dependency>
<dependency>
  <groupId>jakarta.annotation</groupId>
  <artifactId>jakarta.annotation-api</artifactId>
</dependency>
----

> Now we need to generate the actual client for our OpenAPI description, to do so, you should drop the OpenAPI file(in ``yaml`` or ``json`` format) in the ``src/main/openapi`` folder of your project.

Reviewer 3: ```suggestion
Now we need to generate the actual client for our OpenAPI description, to do so, you should drop the OpenAPI file (in ``yaml`` or ``json`` format) in the ``src/main/openapi`` folder of your project.
```
